nixos/modules/services/backup/bupstash.nix

```
@@ -0,0 +1,281 @@
+
+  bupstash = "${cfg.package}/bin/bupstash";
+  defaultUser = "bupstash-server";
+in
+{
+  meta.maintainers = [ lib.maintainers.thubrecht ];
+
+  options.services.bupstash = {
+    package = mkOption {
+      type = package;
+      default = pkgs.bupstash;
+      defaultText = literalExpression "pkgs.bupstash";
+      description = "The bupstash package used for backups.";
+    };
+
+    server = {
+      enable = mkEnableOption "a bupstash server";
+      package = mkOption {
+        type = package;
+        default = pkgs.bupstash;
+        defaultText = literalExpression "pkgs.bupstash";
+        description = "The remote bupstash package used for backups.";
+      };
+      user = mkOption {
+        type = str;
+        default = defaultUser;
+        description = "User for the bupstash server.";
+      };
+      group = mkOption {
+        type = str;
+        default = defaultUser;
+        description = "Group for the bupstash server.";
+      };
+      keys = mkOption {
+        type = listOf str;
+        default = [ ];
+        description = "List of ssh keys with access to the server.";
+      };
+      home = mkOption {
+        type = passwdEntry path;
+        default = "/var/lib/${defaultUser}";
+        description = "The home directory for the bupstash user, wher the backups will be stored.";
+      };
+    };
+
+    jobs = mkOption {
+      type = attrsOf (
+        submodule (
+          { name, config, ... }:
+          {
+            options = {
+              paths = mkOption {
+                type = listOf path;
+                default = [ ];
+                description = "List of paths to backup.";
+              };
+
+              excludes = mkOption {
+                type = listOf str;
+                default = [ ];
+                description = "List of patterns to exlude from the backup.";
+              };
+
+              command = mkOption {
+                type = nullOr str;
+                default = null;
+                description = "Command used for getting the backup data.";
+              };
+
+              key = mkOption {
+                type = path;
+                description = "A path to the key used to backup the data.";
+              };
+
+              user = mkOption {
+                type = str;
+                default = "root";
+                description = "User for the bupstash server.";
+              };
+
+              group = mkOption {
+                type = str;
+                default = "root";
+                description = "Group for the bupstash server.";
+              };
+
+              preStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run before `bupstash put`.";
+              };
+
+              postStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run after `bupstash put`.";
+              };
```

I think these options drive the lingo adoption too hard, because they are executed before and after `borgstash put`. In the context of a put job, these could just be `preHook` and `postHook`, which signifies that the enduser can hook into the execution at defined times.

```suggestion
              preHook = mkOption {
                type = lines;
                default = "";
                description = "Commands to run before `bupstash put`.";
              };

              postHook = mkOption {
                type = lines;
                default = "";
                description = "Commands to run after `bupstash put`.";
              };
```